### PR TITLE
Add attestation stubs

### PR DESCRIPTION
**Pull request by backmeupplz**
Added attestations for:
* Founders/vc (via knowledge of token)
  * `1` type — founders
  * `2` type — VCs
* YC (`0` type) via knowledge of:
  * Email attestation
  * Balance attestation (OrangeDAO)
  * Twitter attestation

### CODE REVIEW

**Comment by MixailE** [comment]
[Add email attestation for yc users](https://trello.com/c/T4LwabMo/2754-add-email-attestation-for-yc-users)

**Comment by MixailE** [comment]
[Add twitter attestation for yc users](https://trello.com/c/HSoKsOk2/2755-add-twitter-attestation-for-yc-users)

File: src/controllers/verify-yc.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    const eddsaSignature = await eddsaSigPoseidon([
+      0,
+      ...utils.toUtf8Bytes(emailHash),
+    ])
+
+    const domain = email.split('@')[1].toLowerCase()
+
+    // TODO:
+    // 3. Send email with signature (must look nice and ketl branded)
+    void sendEmail({
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:38]: Feels like we should extract procedures before into separate helper function like `encryptAndSignEmail`

Comment (T-Damer) [reply src/controllers/verify-yc.ts:38]: The controller should be as simple as possible

File: src/controllers/verify-yc.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    {
+      message,
+      ownerAddress,
+      signature,
+      threshold,
+      tokenAddress = zeroAddress,
+      tokenId,
+    }: BalanceUniqueVerifyBody & AddressVerifyBody
+  ) {
+    // 1. Verify signature and that ownerAddress signed the signature
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:88]: We should encapsulate logic into helper functions with appropriate names to get rid of those comments

File: src/controllers/verify-yc.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    const balanceHash = await poseidonHash(eddsaMessage)
+
+    // 4. Sign `[0, balanceHash]` with EdDSA, (0 = yc attestation)
+    const eddsaSignature = await eddsaSigPoseidon([
+      0,
+      ...utils.toUtf8Bytes(balanceHash),
+    ])
+
+    // 5. Return signature
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:128]: Yet again, we should extract all logic above to make controller look simpler

File: src/controllers/verify-yc.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  @Post('/email-unique')
+  async sendUniqueEmail(
+    @Body({ required: true }) { email }: EmailUniqueVerifyBody
+  ) {
+    // 1. Create `emailHash = poseidon([0, email])`, (0 = email attestation)
+    const eddsaMessage = ['0', utils.hexlify(utils.toUtf8Bytes(email))]
+
+    const emailHash = await poseidonHash(eddsaMessage)
+
+    // 2. Sign `[0, emailHash]` with EdDSA, (0 = yc attestation)
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:28]: Maybe we should create `enum` with proper names so we don't need to write what is `0` every time?

File: src/controllers/verify-yc.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    @Ctx() ctx: Context,
+    @Body({ required: true }) { token }: TwitterBody
+  ) {
+    // 1. Verify token
+    let userId = ''
+    try {
+      const { id } = await fetchUserProfile(token)
+      userId = id
+    } catch (e) {
+      console.log(e)
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:57]: Probably `console.error`, but actually shouldn't differ

File: src/helpers/signatures/eddsaSigPoseidon.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { BigNumber, utils } from 'ethers'
+import { buildEddsa, buildPoseidon } from 'circomlibjs'
+import env from '@/helpers/env'
+
+const privateKey = utils.arrayify(env.EDDSA_PRIVATE_KEY)
+
+export default async function (message: (number | BigNumber)[] | Uint8Array) {
+  // Message
+  const poseidon = await buildPoseidon()
```
Comment (T-Damer) [inline src/helpers/signatures/eddsaSigPoseidon.ts:9]: Should we actually build those each time? Maybe we can cache those guys into `const`?

File: src/helpers/signatures/eddsaSigPoseidon.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { BigNumber, utils } from 'ethers'
+import { buildEddsa, buildPoseidon } from 'circomlibjs'
+import env from '@/helpers/env'
+
+const privateKey = utils.arrayify(env.EDDSA_PRIVATE_KEY)
+
+export default async function (message: (number | BigNumber)[] | Uint8Array) {
+  // Message
```
Comment (T-Damer) [inline src/helpers/signatures/eddsaSigPoseidon.ts:8]: I think we don't need this comment

File: src/helpers/getBalance.ts [hunk 8]
```diff
@@ -3,12 +3,19 @@ import { ethers, providers } from 'ethers'
...
+  provider: providers.Provider,
   ownerAddress: string,
-  provider: providers.Provider
+  tokenAddress = zeroAddress,
+  tokenId?: number
 ) {
   if (tokenAddress === zeroAddress) {
     return provider.getBalance(ownerAddress)
+  } else if (tokenId) {
+    const abi = [
```
[the remark this answers is not in the document]
Comment (MixailE) [reply src/helpers/getBalance.ts:14]: This is a universal abi, not only for our contracts

File: src/controllers/verify-yc.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        polygonProvider,
+        ownerAddress,
+        tokenAddress,
+        tokenId
+      )
+      if (balance.lt(threshold)) {
+        return ctx.throw(badRequest('Not enough balance'))
+      }
+    } catch {
+      return ctx.throw(badRequest("Can't fetch the balances"))
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:87]: Should we log something?

File: src/controllers/verify-yc.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      tokenAddress = zeroAddress,
+      tokenId,
+    }: BalanceUniqueVerifyBody & AddressVerifyBody
+  ) {
+    const signerAddress = ethers.utils
+      .verifyMessage(message, signature)
+      .toLowerCase()
+
+    if (signerAddress !== ownerAddress) {
+      return ctx.throw(badRequest('Invalid ownerAddress'))
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:73]: Should we trow this like plain `new Error` and do:
```TS
catch (e) { 
ctx.throw(badRequest(e.message || 'Not enough balance') 
}
```

File: src/controllers/verify-yc.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+@Controller('/verify-yc')
+export default class VerifyYCController {
+  @Post('/email-unique')
+  async sendUniqueEmail(
+    @Body({ required: true }) { email }: EmailUniqueVerifyBody
+  ) {
+    const { message, signature } = await signAttestationMessage(
+      Attestation.YC,
+      Verification.email,
+      utils.hexlify(utils.toUtf8Bytes(email))
```
Comment (T-Damer) [inline src/controllers/verify-yc.ts:27]: We're doing `utils.hexlify(utils.toUtf8Bytes(...))` in many places, should we extract it into `helpers`?

File: src/helpers/getBalance.ts [hunk 12]
```diff
@@ -3,12 +3,19 @@ import { ethers, providers } from 'ethers'
...
-  tokenAddress: string,
+  provider: providers.Provider,
   ownerAddress: string,
-  provider: providers.Provider
+  tokenAddress = zeroAddress,
+  tokenId?: number
 ) {
   if (tokenAddress === zeroAddress) {
     return provider.getBalance(ownerAddress)
+  } else if (tokenId) {
```
Comment (T-Damer) [inline src/helpers/getBalance.ts:13]: Should we check `tokenId !== undefined` here? Can it be `0`?

File: src/models/Verification.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
+enum Verification {
```
Comment (T-Damer) [inline src/models/Verification.ts:1]: Nice, nice, those enums are perfect

File: src/helpers/env.ts [hunk 14]
```diff
@@ -15,8 +16,10 @@ export default cleanEnv(process.env, {
   ECDSA_PRIVATE_KEY: str(),
   EDDSA_PRIVATE_KEY: str(),
   ETH_NETWORK: str({ default: ETH_NETWORK }),
+  ETH_POLYGON_NETWORK: str({ default: ETH_MUMBAI_NETWORK }),
   ETH_RPC: str({ default: ETH_RPC }),
   ETH_RPC_MAINNET: str({ default: ETH_RPC_MAINNET }),
+  ETH_RPC_POLYGON: str(),
```
Comment (smyzin) [inline src/helpers/env.ts:22]: Should we set here the default value as in `ETH_RPC_MAINNET`?

Comment (MixailE) [reply src/helpers/env.ts:22]: We don't have default RPC for Mumbai

File: src/helpers/getBalance.ts (hunk 12 above)
Comment (MixailE) [reply src/helpers/getBalance.ts:13]: Year, you right

File: src/controllers/verify-yc.ts (hunk 10 above)
Comment (MixailE) [reply src/controllers/verify-yc.ts:75]: I think we can refactor it later

File: src/controllers/verify-yc.ts [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+    try {
+      const balance = await getBalance(
+        polygonProvider,
+        ownerAddress,
+        tokenAddress,
+        tokenId
+      )
+      if (balance.lt(threshold)) {
+        return ctx.throw(badRequest('Not enough balance'))
```
[the remark this answers is not in the document]
Comment (MixailE) [reply src/controllers/verify-yc.ts:86]: I think we can refactor it later

File: src/controllers/verify-yc.ts (hunk 9 above)
Comment (MixailE) [reply src/controllers/verify-yc.ts:90]: Fixed

**Review by T-Damer — APPROVED** [review]

**Review by smyzin — APPROVED** [review]

**Pull request merged**